File: 250320/강력한 폭발/strong-explosion.cpp
#include <bits/stdc++.h>
using namespace std;
#define MAX_N 20
#define BOMB_TYPE 3

int n;
int grid[MAX_N][MAX_N];
bool bombed[MAX_N][MAX_N];

int ans;
vector<pair<int, int>> bomb_pos;

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n;
}

void Bomb(int x, int y, int k) {
    pair<int, int> bomb_shapes[BOMB_TYPE+1][5] = {
        {},
        {{-2, 0}, {-1, 0}, {0, 0}, {1, 0}, {2, 0}},
        {{-1, 0}, {1, 0}, {0, 0}, {0, -1}, {0, 1}},
        {{-1, -1}, {-1, 1}, {0, 0}, {1, -1}, {1, 1}}
    };

    for(int i=0; i<5; i++) {
        int dx, dy;
        tie(dx, dy) = bomb_shapes[k][i];

        int nx = x+dx, ny = y+dy;
        if(InRange(nx, ny))
            bombed[nx][ny] = true;
    }
}

int Calc() {
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            bombed[i][j] = false;
        }
    }

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            if(grid[i][j]) {
                Bomb(i, j, grid[i][j]);
            }
        }
    }

    int cnt = 0;
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            if(bombed[i][j]) cnt++;
        }
    }

    return cnt;
}

void FindMaxArea(int idx) {
    if(idx == (int)bomb_pos.size()) {
        ans = max(ans, Calc());
        return;
    }

    for(int i=1; i<=3; i++) {
        int x, y;
        tie(x, y) = bomb_pos[idx];

        grid[x][y] = i;
        FindMaxArea(idx+1);
        grid[x][y] = 0;
    }
}

int main() {
    cin >> n;

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> grid[i][j];
            if(grid[i][j]) bomb_pos.push_back(make_pair(i, j));
        }
    }

    FindMaxArea(0);

    cout << ans;
    return 0;
}

File: 250611/강력한 폭발/strong-explosion.py
n = int(input())

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

bombed = [
    [False for _ in range(n)]
    for _ in range(n)
]

bomb_type = [
    [],
    [[-2, 0], [-1, 0], [0, 0], [1, 0], [2, 0]],
    [[-1, 0], [1, 0], [0, 0], [0, -1], [0, 1]],
    [[-1, -1], [-1, 1], [0, 0], [1, -1], [1, 1]]
]

bomb_pos = []

ans = 0

def inrange(x, y):
    return 0<=x<n and 0<=y<n

def bomb(x, y, bt):
    for i in range(5):
        dx, dy = bomb_type[bt][i]
        nx, ny = x+dx, y+dy

        if inrange(nx, ny):
            bombed[nx][ny] = True

def calc():
    for i in range(n):
        for j in range(n):
            bombed[i][j] = False
    
    for i in range(n):
        for j in range(n):
            if grid[i][j]:
                bomb(i, j, grid[i][j])
    
    return sum(
        bombed[i][j]
        for i in range(n)
        for j in range(n)
    )

def find_max_area(cnt):
    global ans

    if cnt == len(bomb_pos):
        ans = max(ans, calc())
        return

    for i in range(1, 4):
        x, y = bomb_pos[cnt]

        grid[x][y] = i
        find_max_area(cnt+1)
        grid[x][y] = 0

for i in range(n):
    for j in range(n):
        if grid[i][j] == 1:
            bomb_pos.append((i, j))


find_max_area(0)

print(ans)


